### PR TITLE
Dispose layer tree node event listeners.

### PR DESCRIPTION
**Pull request by bentrm**
I think there are some dangling event listeners associated with the layer node. I hope to fix this with this PR.

As far as I can tell, there is no reasonable way to hook into the destruction of the LayerTreeModel of the LayerNode to fix this so I used the maps 'removelayer' event. 'preremovelayer' is not an option as other listeners outside of this scope may cancel removing the layer. That's also the reason for caching a reference to the actual map. When 'removelayer' is called the layers map property is already lost.

### CODE REVIEW

**Comment by marcjansen** [comment]
Any chance you can add some tests? On first glance this looks fine.

**Comment by bentrm** [comment]
There are no explicit tests for GeoExt.tree.LayerNode right now so I was a bit hesitant and deferred it. I'm happy to help but will not have time to look into this before the weekend.

File: src/GeoExt/tree/LayerNode.js [hunk 1]
```diff
@@ -115,6 +136,35 @@ Ext.define('GeoExt.tree.LayerNode', {
...
     /**
+     * Disposes event handlers that have been added during initialization of plugin.
+     *
+     * @private
+     */
+    onMapRemovelayer: function(evt) {
+        var target = this.target,
+            layer = target.get('layer')
+
+        if (evt.layer !== layer) return;
```
Comment (marcjansen) [inline src/GeoExt/tree/LayerNode.js]: Please also use curly braces around one liners.

File: src/GeoExt/tree/LayerNode.js [hunk 2]
```diff
@@ -115,6 +136,35 @@ Ext.define('GeoExt.tree.LayerNode', {
     },
 
     /**
+     * Disposes event handlers that have been added during initialization of plugin.
+     *
+     * @private
+     */
+    onMapRemovelayer: function(evt) {
+        var target = this.target,
+            layer = target.get('layer')
```
Comment (marcjansen) [inline src/GeoExt/tree/LayerNode.js]: Missing **`;`**

**Comment by marcjansen** [comment]
> As far as I can tell, there is no reasonable way to hook into the destruction of the LayerTreeModel of the LayerNode to fix this …

Have you tried (untested):

```javascript
// …
init: function(target) {
    this.target = target;
    this.target.addListener('beforedestroy', this.cleanupMethod, this);
    // …
}
// …
```

Except for this and two other comments this is good to go.

**Comment by bentrm** [comment]
Yes, I tried that. `this.target` is an instance of GeoExt.data.LayerTreeModel which is an Ext.data.Model. The only event that is advertised seems to be `idchanged`. Furthermore, I tried to destroy the LayerNode plugin by hooking into the destroy method of the `LayerTreeModel` but it is never called as "destroy" of Ext.data.Model is used to delete the record which seems not to be the case here.
I think the proposed solution is borderline hackish but I'm out of ideas.

**Comment by marcjansen** [comment]
I will give my OK for a merge if we add a big TODO somewhere and maybe even an issue as the approach seems more than hackish to me. 

Nonetheless the result is better than what we had before, so plus one for merging this once it is updated with a big TODO.

Can ypu update this @bentrm.

**Comment by bentrm** [comment]
> as the approach seems more than hackish to me.

I agree, absolutely. Will add a TODO.

**Comment by chrismayer** [comment]
Hi guys,
any news on this? thanks!

**Comment by bentrm** [comment]
Promise to take care of this this afternoon. Sorry about that, I know you want to ship on 4th. :shipit:

**Comment by bentrm** [comment]
I amended the last commit to include TODOs hope you're ok with that. Thanks for your patience.

**Comment by marcjansen** [comment]
:+1:  thanks!

**Pull request merged**